File: Core/DotNetCore/MediaHatchCore/DownloadManagerWatchFolder.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace MediaHatchCore
{
	public class DownloadManagerWatchFolder : DownloadManager
	{
		private const string _defaultOptionsFile = "dlmanager.options";
		private const string _defaultLinksFileExtension = ".crawljob";
		private const string _linksFileAddedFolderName = "added";

		private const string _mediaTypeWatchFolderSeparator = ";";

		public DownloadManagerWatchFolder()
		{
			string[] options = File.ReadAllLines(_defaultOptionsFile);
			foreach (string option in options)
			{
				if (option.Contains(_mediaTypeWatchFolderSeparator))
				{
					string[] mediaTypeFolderPaths = option.Split(new string[] { _mediaTypeWatchFolderSeparator }, StringSplitOptions.RemoveEmptyEntries);
					if (mediaTypeFolderPaths.Length > 0)
					{
						Type t = Type.GetType(mediaTypeFolderPaths[0]);
						if (t != null)
						{
							_mediaTypeToWatchFolderPath.Add(t, mediaTypeFolderPaths.Length > 1 ? mediaTypeFolderPaths[1] : string.Empty);
							_mediaTypeToDownloadFolderPath.Add(t, mediaTypeFolderPaths.Length > 2 ? mediaTypeFolderPaths[2] : string.Empty);
						}
					}
				}
				else if (option == "downloadFolderPerTvShow")
					_downloadFolderPerTvShow = true;
				else
					_orderedFavoriteFileHosts.Add(option);
			}
		}

		public void Download(List<Media> media)
		{
			foreach (Media currentMedia in media)
				Download(currentMedia);
		}

		public override void Download(Media media)
		{
			if (media.LinksPackageAvailable)
				InternalDownload(media);
		}

		public List<string> GetProcessedMediaSupportedFullNames()
		{
			List<string> processedMediaFullNames = new List<string>();
			foreach (string folder in _mediaTypeToWatchFolderPath.Values)
			{
				processedMediaFullNames.AddRange(Directory.EnumerateFiles(folder, $"*{_defaultLinksFileExtension}").Select(s => Path.GetFileNameWithoutExtension(s)));
				processedMediaFullNames.AddRange(Directory.EnumerateFiles(Path.Combine(folder, _linksFileAddedFolderName), $"*{_defaultLinksFileExtension}").Select(s => Path.GetFileNameWithoutExtension(s)));
			}
			return new List<string>(processedMediaFullNames.Distinct());
		}

		private void InternalDownload(Media media)
		{
			string watchFolderPath;
			if (_mediaTypeToWatchFolderPath.TryGetValue(media.GetType(), out watchFolderPath))
			{
                if (Directory.Exists(watchFolderPath))
                {
                    string filePathSupportedMediaFullName = media.FilePathSupportedFullName();
                    string filePath = Path.Combine(watchFolderPath, $"{filePathSupportedMediaFullName}{_defaultLinksFileExtension}");
                    string addedFilePath = Path.Combine(watchFolderPath, _linksFileAddedFolderName, $"{filePathSupportedMediaFullName}{_defaultLinksFileExtension}");
                    if (!File.Exists(addedFilePath))
                    {
                        string crawlJobContent = $"autoStart=TRUE\nautoConfirm=TRUE\ntext={media.GetFirstFavoriteLinksPackage(_orderedFavoriteFileHosts).ConcatenatedLinks}\npackageName={filePathSupportedMediaFullName}\ncomment={filePathSupportedMediaFullName}";
                        string downloadFolderPath;
                        if (_mediaTypeToDownloadFolderPath.TryGetValue(media.GetType(), out downloadFolderPath))
                        {
                            if (_downloadFolderPerTvShow && media is TvShowEpisode)
                                downloadFolderPath = Path.Combine(downloadFolderPath, media.FilePathSupportedName());
                            crawlJobContent += $"\ndownloadFolder={downloadFolderPath}";
                        }
                        File.WriteAllText(filePath, crawlJobContent);
                    }
                }
			}
		}

		private List<string> _orderedFavoriteFileHosts = new List<string>();
		private Dictionary<Type, string> _mediaTypeToWatchFolderPath = new Dictionary<Type, string>();
		private Dictionary<Type, string> _mediaTypeToDownloadFolderPath = new Dictionary<Type, string>();
		private bool _downloadFolderPerTvShow;
	}
}

File: Backend/media-hatch-server/src/services/download-manager/download-manager-watch-folder.ts
import { DownloadManager } from './download-manager';
import { Media, TvShowEpisode } from 'media-hatch-core';
import * as fs from 'fs-extra';
import * as path from 'path';
import { Factories } from '../utils/factories';
import { FileNameSanitizer } from '../utils/file-name-sanitizer';

export class DownloadManagerWatchFolder extends DownloadManager {
	private static readonly defaultOptionsFile = 'dlmanager.options';
	private static readonly defaultLinksFileExtension = '.crawljob';
	private static readonly linksFileAddedFolderName = 'added';

	private static readonly mediaTypeWatchFolderSeparator = ';';

	private orderedFavoriteFileHosts = new Array<string>();
	private mediaFactoryToWatchFolderPath = new Map<new () => object, string>();
	private mediaFactoryToDownloadFolderPath = new Map<new () => object, string>();
	private downloadFolderPerTvShow = false;

	constructor() {
		super();
		this.initializeOptions();
	}

	private async initializeOptions() {
		const fileOptions = (await fs.readFile(DownloadManagerWatchFolder.defaultOptionsFile, 'utf8')).replace(/^\ufeff/, ''); // removes BOM if present
		const options = fileOptions.split(/\r?\n/);
		for (const option of options) {
			if (option.includes(DownloadManagerWatchFolder.mediaTypeWatchFolderSeparator)) {
				const mediaTypeFolderPaths = option.split(DownloadManagerWatchFolder.mediaTypeWatchFolderSeparator);
				if (mediaTypeFolderPaths.length > 0) {
					const factory = Factories.getFactory(mediaTypeFolderPaths[0]);
					if (factory != null) {
						this.mediaFactoryToWatchFolderPath.set(factory, mediaTypeFolderPaths.length > 1 ? mediaTypeFolderPaths[1] : '');
						this.mediaFactoryToDownloadFolderPath.set(factory, mediaTypeFolderPaths.length > 2 ? mediaTypeFolderPaths[2] : '');
					}
				}
			} else if (option === 'downloadFolderPerTvShow') {
				this.downloadFolderPerTvShow = true;
			} else {
				this.orderedFavoriteFileHosts.push(option);
			}
		}
	}

	public async download(media: Media): Promise<void> {
		if (media.linksPackageAvailable) {
			await this.internalDownload(media);
		}
	}

	public async getProcessedMediaSupportedFullNames(): Promise<string[]> {
		const processedMediaFullNames = new Array<string>();
		for (const folder of this.mediaFactoryToWatchFolderPath.values()) {
			processedMediaFullNames.push(
				...(await fs.readdir(folder))
					.filter(s => s.endsWith(DownloadManagerWatchFolder.defaultLinksFileExtension))
					.map(s => path.basename(s, path.extname(s)))
			);
			processedMediaFullNames.push(
				...(await fs.readdir(path.join(folder, DownloadManagerWatchFolder.linksFileAddedFolderName)))
					.filter(s => s.endsWith(DownloadManagerWatchFolder.defaultLinksFileExtension))
					.map(s => path.basename(s, path.extname(s)))
			);
		}
		return [...new Set(processedMediaFullNames)];
	}

	private async internalDownload(media: Media): Promise<void> {
		const mediaFactory = Factories.getFactory(media.type);
		if (this.mediaFactoryToWatchFolderPath.has(mediaFactory)) {
			const watchFolderPath = this.mediaFactoryToWatchFolderPath.get(mediaFactory);
			if (await fs.pathExists(watchFolderPath)) {
				const filePathSupportedMediaFullName = FileNameSanitizer.sanitizeFullName(media);
				const fileName = filePathSupportedMediaFullName + DownloadManagerWatchFolder.defaultLinksFileExtension;
				const filePath = path.join(watchFolderPath, fileName);
				const addedFilePath = path.join(watchFolderPath, DownloadManagerWatchFolder.linksFileAddedFolderName, fileName);
				if (!fs.pathExists(addedFilePath)) {
					let crawlJobContent = `autoStart=TRUE\nautoConfirm=TRUE\ntext=${
						media.getFirstFavoriteLinksPackage(this.orderedFavoriteFileHosts).concatenatedLinks
					}\npackageName=${filePathSupportedMediaFullName}\ncomment=${filePathSupportedMediaFullName}`;
					if (this.mediaFactoryToDownloadFolderPath.has(mediaFactory)) {
						let downloadFolderPath = this.mediaFactoryToDownloadFolderPath.get(mediaFactory);
						if (this.downloadFolderPerTvShow && media instanceof TvShowEpisode) {
							downloadFolderPath = path.join(downloadFolderPath, FileNameSanitizer.sanitizeName(media));
						}
						crawlJobContent += `\ndownloadFolder=${downloadFolderPath}`;
					}
					await fs.writeFile(filePath, crawlJobContent);
				}
			}
		}
	}
}
